### PR TITLE
Bookmarking enhancements

### LINKED ISSUE #826: Duplicate bookmark prevention
## Description

When a user has already bookmarked a message, we check if the same message is already bookmarked and warn the user regarding the same

## Reasoning
This feature would help detect duplicate bookmarks 

## Proposed Implementation
No ideas as if now :)

## Would you like to implement this yourself?

- [ ] I'd like to implement this feature myself
- [x] Anyone can implement this feature

### LINKED ISSUE #918: un bookmark command
## Description
There would be a way to remove a bookmark. This would be useful while mistakenly creating a bookmark or if you have many of bookmarks and don't want to have them there anymore. 
## Reasoning

If you happen to create a bookmark for a wrong message, you could delete that bookmark. Or if you just have a old bookmark that you don't anymore need. It would keep the bookmarks clear.

## Proposed Implementation

Possibly creating a tag for each bookmark, as example `#4321`. With that you could do `.unbm 4321` or `.unbm #4321`

## Would you like to implement this yourself?

- [ ] I'd like to implement this feature myself
- [x] Anyone can implement this feature

### PR DESCRIPTION
**Pull request by ChrisLovering**
## Relevant Issues

Closes #826
Closes #918

## Description

The bot now tracks ids of messages that the user has bookmarked in Redis, so that a user can't unintentionally bookmark the same message more than once.

This also adds the ability for users to delete messages in DMs with the bot via the use of the `bookmark delete` command.

On top of this I have done some minor refactoring and fixed a bug that could rase an AtrributeError, that we haven't yet encountered.

## Did you:

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] Read all the comments in this template?
- [x] Ensure there is an issue open, or link relevant discord discussions?
- [x] Read and agree to the [contributing guidelines](https://pythondiscord.com/pages/contributing/contributing-guidelines/)?

### CODE REVIEW

File: bot/exts/utilities/bookmark.py
Comment (hedyhli) [inline bot/exts/utilities/bookmark.py:141]: I think that bookmark refers to a "pin" or a reference in the user's DMs to a target message. i.e.: the bookmark is the DM itself, which is *just* getting created in here.
```suggestion
        Send the author a link to the specified message via DMs.
```

File: bot/exts/utilities/bookmark.py
Comment (hedyhli) [inline bot/exts/utilities/bookmark.py:219]: I'm not sure how it can be implemented, but it would be a nice addition to allow users to use the target bookmark message to delete. For example, they use `bm <message-id>` to bookmark, realized they bookmarked the wrong thing. So they go to their DMs and run `bm del <message-id>` to delete it. Here `<message-id>` would only need to be copied once and reused when deleting. Sure, they can just use the reply-to to delete, but making it like I said above would make it a sort of symmetric operation -- using the same argument to add and delete.

**Review by hedyhli — CHANGES_REQUESTED** [review]
Haven't tested yet

File: bot/exts/utilities/bookmark.py
Comment (Shivansh-007) [reply bot/exts/utilities/bookmark.py:204]: The extra `,` was present so that `("del")` remains a tuple, otherwise, python would consider it as a single string. Trailing commas are only used for multiline tuples or single length tuples IIRC.

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py:39]: Yea that's a good spot. I'll change this, but I want to make it clear that this is serialised, so will change it slightly more than suggested.

Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py]: https://github.com/python-discord/sir-lancebot/pull/929/commits/fbdfae42a40df0242d8050438d8c99aea193d386

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py]: https://github.com/python-discord/sir-lancebot/pull/929/commits/fbdfae42a40df0242d8050438d8c99aea193d386

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py]: forcepushed onto https://github.com/python-discord/sir-lancebot/pull/929/commits/6fcefb0a6b2153731b72d7ce194531170c5fcd3e

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py]: dropped the commit that added this and forcepushed

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py]: https://github.com/python-discord/sir-lancebot/pull/929/commits/fbdfae42a40df0242d8050438d8c99aea193d386
Also changed all instances of user to member.

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py]: https://github.com/python-discord/sir-lancebot/pull/929/commits/fbdfae42a40df0242d8050438d8c99aea193d386

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py]: https://github.com/python-discord/sir-lancebot/pull/929/commits/f8d09b2eabbfca8829312f4237a953d6d890caa0

Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py]: https://github.com/python-discord/sir-lancebot/pull/929/commits/9c1289af534b9d33d1f38ae5dff0edb2208551da improved var names

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py:223]: Yea, that one would require use to store a lookup from message ID to dm message ID for each member, possibly over complicating it. I may be tempted to say we just keep the interface semi-restricted like this instead.

File: bot/exts/utilities/bookmark.py
Comment (hedyhli) [reply bot/exts/utilities/bookmark.py]: Longer but more descriptive, I like it! Might wanna resolve this?

Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py]: Nice, thanks :D Yea, just wanted to wait for your feedback to make sure.

**Review by hedyhli — CHANGES_REQUESTED** [review]
Tested locally, works well 👍 Just several clarifications on its behavior and some trivial fixes.

File: bot/exts/utilities/bookmark.py
Comment (onerandomusername) [reply bot/exts/utilities/bookmark.py:225]: *technically* this could be hacked through. 

As far as I can tell, the embed title has been "Bookmark" for all bookmarks sent in dms.

Comment (Nipa-Code) [reply bot/exts/utilities/bookmark.py:225]: imo it would not be bad if it allows other messages to be deleted as well. Also there wouldn't be need for storing bookmarks anywhere in that case. The original idea was for bookmarks but this does not sound bad idea if it allows to delete other messages.

Comment (hedyhli) [reply bot/exts/utilities/bookmark.py:225]: Yeah it isn't a bad idea, but IMO it's a bit of a hack and weird to make sir-lancebot delete any of its messages using a command called `.unbm`

Comment (Nipa-Code) [reply bot/exts/utilities/bookmark.py:225]: Ye, the `.unbm` doesn't fit there in that case. If it would be allowed to delete any messages the `.unbm` would be something like `.del`/`.delete` that should be existing.

File: bot/exts/utilities/bookmark.py
Comment (Bluenix2) [inline bot/exts/utilities/bookmark.py:25]: In the same channel right?

File: bot/exts/utilities/bookmark.py
Comment (Bluenix2) [inline bot/exts/utilities/bookmark.py:26]:
```suggestion
    "\n3. Lookup by a message URL"
```

Comment (Bluenix2) [inline bot/exts/utilities/bookmark.py:26]: Do we want to add punctuation at the end?

File: bot/exts/utilities/bookmark.py
Comment (Bluenix2) [reply bot/exts/utilities/bookmark.py:60]: Yes this should be removed, since it won't work for appending.

File: bot/exts/utilities/bookmark.py
Comment (Bluenix2) [inline bot/exts/utilities/bookmark.py:63]: Since the previous if-statement should be removed, this can now raise a ValueError.

Comment (hedyhli) [reply bot/exts/utilities/bookmark.py:63]: For `list.remove`? The if statement can be kept for `add_or_remove == "remove"` (even though it may be guarded by [this if `.unbm` is only for actual bookmarks](https://github.com/python-discord/sir-lancebot/pull/929#discussion_r743390980)).

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py:26]: This would make it inconsistent with the other lines.

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py:25]: Nope, that's not how it works in 2.0 now.

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py:223]: I was going to say this wouldn't be easily possible since asyncredis cache doesn't really support these types, that's why I'm serialising. but then I realised if I'm already serialising, then I can just serialise dicts instead. I will rethink the approach for this.

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py:207]: Yea, this is an d.py thing. I could remove the check here and do it inside the command instead, that way we can have customer handling.

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py:150]: Yea, I will making this interface better with the custom check. Will resolve this convo and keep the discussion in the other one.

File: bot/exts/utilities/bookmark.py
Comment (ChrisLovering) [reply bot/exts/utilities/bookmark.py:225]: I chose not to limit this to only message IDs in redis as then users wouldn't be able to delete messages bookmarked before this PR was merged.

I think root-aliasing this as .delete could be good, especially with the customer dm check I will add which will explain the interface if ran in-server.

File: bot/exts/utilities/bookmark.py
Comment (onerandomusername) [inline bot/exts/utilities/bookmark.py:110]: Seems like this will error if the user cannot be dmed, and won't handle it safely.

Comment (onerandomusername) [reply bot/exts/utilities/bookmark.py:110]: Ah, it catches the error below. Perhaps update the docstring, since it clarifies one of the issues it will raise but not all.

File: bot/exts/utilities/bookmark.py
Comment (onerandomusername) [reply bot/exts/utilities/bookmark.py:129]: I like this function, however it seems to be two seperate messages. Discord can now support up to 10 embeds per message, so if this does occur they should be in the same message IMO.

File: bot/exts/utilities/bookmark.py
Comment (onerandomusername) [inline bot/exts/utilities/bookmark.py:160]: Perhaps it shouldn't even disclose that the channel exists?

File: bot/exts/utilities/bookmark.py
Comment (onerandomusername) [reply bot/exts/utilities/bookmark.py:207]: I understand. I don't think the UI would make too much sense if a bookmark cannot be deleted from dms.

I feel like it would make the most sense to take a reply or a message as a parameter. if a user uses `.bm del` and replying to a bookmark, it should delete it, just like `.bm` can be used to bookmark a reply.

File: bot/exts/utilities/bookmark.py
Comment (onerandomusername) [reply bot/exts/utilities/bookmark.py:63]: Rather than use strings at all, what about an enum of `Action.DELETE` and `Action.ADD` or something similar?

File: bot/exts/utilities/bookmark.py
Comment (onerandomusername) [inline bot/exts/utilities/bookmark.py:85]: 'message' here feels like a discord message, but its just a string.

This may be worth raising another issue given how many different methods are like this in the bot. ERROR_REPLIES and Colours.soft_red make a lot of error embeds, that it might be worth putting this in a utils file now and standardizing them as they are come across.

File: bot/exts/utilities/bookmark.py
Comment (onerandomusername) [inline bot/exts/utilities/bookmark.py:103]: Also `Raises error if the user cannot be messaged.`

**Review by onerandomusername — CHANGES_REQUESTED** [review]

**Comment by ChrisLovering** [comment]
i will re-make this PR, as a lot of the context in this PR is irrelevant due to the new direction.

**Pull request closed without merging**
